ndk-build/src/manifest.rs

```diff
@@ -1,258 +1,345 @@
+                let mut f = features.clone();
+                f.push(Feature {
+                    name: None,
+                    required: Some(true),
+                    opengles_version: Some((3, 1)),
+                });
+                return serializer.serialize_some(&f);
+            }
+            serializer.serialize_some(features)
+        }
+        None => serializer.serialize_none(),
     }
+}
 
+impl AndroidManifest {
     pub fn write_to(&self, dir: &Path) -> Result<(), NdkError> {
         let mut file = File::create(dir.join("AndroidManifest.xml"))?;
-        writeln!(file, "{}", self.to_string())?;
+        let serialized = quick_xml::se::to_string(&self).unwrap();
+        writeln!(file, r#"<?xml version="1.0" encoding="utf-8"?>"#)?;
+        writeln!(file, "{}", serialized)?;
         Ok(())
     }
 }
 
-#[derive(Debug)]
-pub struct Feature {
-    pub name: String,
-    pub required: bool,
+/// See https://developer.android.com/guide/topics/manifest/application-element
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct Application {
+    #[serde(rename(serialize = "android:debuggable"))]
+    pub debuggable: Option<bool>,
+    #[serde(rename(serialize = "android:theme"))]
+    pub theme: Option<String>,
+    #[serde(rename(serialize = "android:hasCode"))]
+    #[serde(default = "default_has_code")]
+    pub has_code: Option<bool>,
+    #[serde(rename(serialize = "android:icon"))]
+    pub icon: Option<String>,
+    #[serde(rename(serialize = "android:label"))]
+    #[serde(default)]
+    pub label: String,
+
+    #[serde(rename(serialize = "meta-data", deserialize = "metadata"))]
+    pub meta_datas: Option<Vec<MetaData>>,
+    #[serde(default)]
+    pub activity: Activity,
 }
 
-impl Feature {
-    pub fn to_string(&self) -> String {
-        format!(
-            r#"<uses-feature android:name="{}" android:required="{}"/>"#,
-            &self.name, self.required,
-        )
+impl Default for Application {
+    fn default() -> Self {
+        Self {
+            debuggable: None,
+            theme: None,
+            has_code: default_has_code(),
+            icon: None,
+            label: Default::default(),
+            meta_datas: None,
+            activity: Default::default(),
+        }
     }
 }
 
-#[derive(Debug)]
-pub struct ApplicationMetadata {
-    pub name: String,
-    pub value: String,
+/// See https://developer.android.com/guide/topics/manifest/activity-element
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct Activity {
+    #[serde(rename(serialize = "android:configChanges"))]
+    #[serde(default = "default_config_changes")]
+    pub config_changes: Option<String>,
+    #[serde(rename(serialize = "android:label"))]
+    pub label: Option<String>,
+    #[serde(rename(serialize = "android:launchMode"))]
+    pub launch_mode: Option<String>,
+    #[serde(rename(serialize = "android:name"))]
+    #[serde(default = "default_activity_name")]
+    pub name: Option<String>,
```

Comment:
Isn't this also always required to be `android.app.NativeActivity` for native activities?

Reply:
The default value is currently always set to `android.app.NativeActivity`. I wasn't sure if in the future support for multiple activities would be implemented so I left it as an `Option` in this PR. Could change that of course to always default to `android.app.NativeActivity`